pip/pip-425.md

```diff
@@ -0,0 +1,127 @@
+        // Default implementation does nothing
+        // Subclass can override this method to implement host availability tracking
+    }
+```
+When resolve one address in serviceUrl, the `ConnectionPool` will invoke `markHostAvailability` to tell availability about that address, so the `ServiceNameResolver` can remove it from `availableAddressPool` if unavailable (or recover back to `availableAddressPool` if available). If all addressed in  `availableAddressPool` are unavailable, the resolver implementation could start getting random addresses from the `allAddressPool`.
+
+For removed addresses, the quarantine duration will increase exponentially. The quarantine duration can be customized via new optional configurations `serviceUrlInitFaultQuarantineDuration` and `serviceUrlMaxFaultQuarantineDuration` in `ClientBuilder`
+That is to say, assuming that `serviceUrlInitFaultQuarantineDuration` is 1 minute and `serviceUrlMaxFaultQuarantineDuration` is 30 minutes, after host1 is judged as unavailable for the first time, it will be isolated for 1 minute and then added to `availableAddressPool` again. If it is judged as unavailable again the next time, it will be isolated for 2 minutes. Similarly, if it is judged as unavailable continuously, the isolation time will be 4 minutes, 8 minutes, 16 minutes, 30 minutes (it reaches serviceUrlMaxFaultQuarantineDuration and no longer increases), 30 minutes... If host1 is determined to be available, the next quarantine duration will be reset to 1 minute.
+
+
+### Usage Example
+```java
+    private void example() throws PulsarClientException {
+        // The failed endpoint will be removed and retried after a period of time, with an initial quarantine duration of 30 seconds increasing exponentially. After reaching the maximum quarantine duration of 1 hour, will be maintained once an hour until a certain succeeds, and the quarantine duration is reset to 0.
+        PulsarClient pulsarClient = PulsarClient.builder()
+                .serviceUrl("pulsar://localhost1:6650,localhost2:6650,localhost3:6650")
+                .serviceUrlInitFaultQuarantineDuration(30000, TimeUnit.MILLISECONDS)
+                .serviceUrlMaxFaultQuarantineDuration(1, TimeUnit.HOURS)
+                .build();    
+
+        // disable removing unavailable serviceUrl endpoints by pass 0 to serviceUrlInitFaultQuarantineDuration and serviceUrlMaxFaultQuarantineDuration
+        PulsarClient pulsarClientDisableReovery = PulsarClient.builder()
+                .serviceUrl("pulsar://localhost1:6650,localhost2:6650,localhost3:6650")
+                .serviceUrlInitFaultQuarantineDuration(0, TimeUnit.MILLISECONDS)
+                .serviceUrlMaxFaultQuarantineDuration(0, TimeUnit.MILLISECONDS)
+                .build();                
+}
+
+```
+## Public-facing Changes
+
+### Public API
+
+1. Add two optional configs `serviceUrlInitFaultQuarantineDuration` and `serviceUrlMaxFaultQuarantineDuration` in `ClientBuilder`
```

Comment:
I don't think that `fault` is needed at all as part of the parameter name. `quarantine` already implies a fault.
`serviceUrlQuarantineInitDuration` and `serviceUrlQuarantineMaxDuration` would be easier to read and understand.